environment/src/main/java/com/sequenceiq/environment/environment/poller/StackPollerProvider.java

```diff
@@ -0,0 +1,83 @@
+package com.sequenceiq.environment.environment.poller;
+
+import com.dyngr.core.AttemptMaker;
+import com.dyngr.core.AttemptResult;
+import com.dyngr.core.AttemptResults;
+import com.dyngr.core.AttemptState;
+import com.sequenceiq.environment.environment.service.stack.StackService;
+import com.sequenceiq.flow.core.FlowConstants;
+import com.sequenceiq.flow.domain.FlowLog;
+import com.sequenceiq.flow.service.flowlog.FlowLogDBService;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.ws.rs.BadRequestException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+@Component
+public class StackPollerProvider {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(StackPollerProvider.class);
+
+    private final StackService stackService;
+
+    private final FlowLogDBService flowLogDBService;
+
+    public StackPollerProvider(
+        StackService stackService,
+        FlowLogDBService flowLogDBService) {
+        this.stackService = stackService;
+        this.flowLogDBService = flowLogDBService;
+    }
+
+    public AttemptMaker<Void> stackUpdateConfigPoller(List<String> stackNames, Long envId, String flowId) {
+        List<String> mutableNamesList = new ArrayList<>(stackNames);
+        return () -> {
+            Optional<FlowLog> flowLog = flowLogDBService.getLastFlowLog(flowId);
+            if (flowLog.isPresent() && flowLog.get().getCurrentState().equals(FlowConstants.CANCELLED_STATE)) {
+                return AttemptResults.finishWith(null);
+            }
+            List<String> remaining = new ArrayList<>();
+            List<AttemptResult<Void>> results = collectStackUpdateConfigResults(mutableNamesList,
+                remaining, envId);
+            mutableNamesList.retainAll(remaining);
+            return evaluateResult(results);
+        };
+    }
+
+    private List<AttemptResult<Void>> collectStackUpdateConfigResults(List<String> stackNames,
+        List<String> remaining, Long envId) {
+        return stackNames.stream()
+            .map(stackName -> fetchStackUpdateConfigResults(remaining, stackName, envId))
+            .collect(Collectors.toList());
+    }
+
+    private AttemptResult<Void> fetchStackUpdateConfigResults(List<String> remainingStacks,
+        String stackName, Long envId) {
+        try {
+            stackService.triggerConfigUpdateForStack(stackName);
+            return AttemptResults.finishWith(null);
+        } catch (BadRequestException e) {
+            remainingStacks.add(stackName);
```

Comment:
so instead of passing in the `remainingStacks` paramater and adding stuff, why don't we return with `AttemptResult<String>`?

Reply:
Because the stack name gets added to the remainingStacks and is processed 
by reference in stackUpdateConfigPoller.  AttemptResult<String>  would not
allow it to be processed in the as I have to have a return type of AttemptResult<Void>
to handle all the cases and force a check against null in the results.

This is the same pattern used for the Datahub polling to stop multiple datahubs.

Reply:
thanks for the clarification